src/Bicep.Core/Emit/PositionTrackingJsonTextWriter.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Bicep.Core.Emit
+{
+    public class PositionTrackingJsonTextWriter : JsonTextWriter
+    {
+        private class PositionTrackingTextWriter : TextWriter
+        {
+            private readonly TextWriter _internalWriter;
+
+            public int CurrentPos;
+
+            public List<int> CommaPositions = new();
+
+            public PositionTrackingTextWriter(TextWriter textWriter)
+            {
+                _internalWriter = textWriter;
+            }
+
+            public override Encoding Encoding => _internalWriter.Encoding;
+
+            public override void Write(char value)
+            {
+                if (value == ',')
```

Comment:
Do we need to handle the case where the comma is in a string?

Reply:
+1. `PositionTrackingJsonTextWriter` is actually overriding `JsonTextWriter` yet we're not taking advantage of any of the method overloads (which have correctly-parsed contextual information about the JSON) - and are instead dropping down to parsing the JSON ourselves inside the `PositionTrackingTextWriter`.

If we're trying to accurately identify positions between JSON elements, wouldn't overriding some of the methods in https://www.newtonsoft.com/json/help/html/t_newtonsoft_json_jsontextwriter.htm be a better option?

Reply:
@shenglol It shouldn't be necessary. This will pick up commas in strings but it shouldn't affect the functionality since the only commas we care about are the structural ones that could affect line numbering. This approach does feel kind of hacky though so per @anthony-c-martin I am going to see if I can clean this up and make it a bit more idiomatic of JSON by overriding the JsonTextWriter methods.

Reply:
So, I took another try at this and I am not able to come up with an implementation that doesn't add more complexity than the current solution I have now. The problem with using overridden JsonTextWriter methods to track mappings is that the 
bicep context is also needed when the mappings are created. However, this context is often not in the same scope where the JSON is being manipulated. There are ways to account for this, but all the ideas I have had so far would just add more overall complexity relative to the current implementation.

If you have some more tangible ideas for a more elegant approach, would be happy to set up some time to talk.

Reply:
> This will pick up commas in strings but it shouldn't affect the functionality since the only commas we care about are the structural ones that could affect line numbering.

That may not be true since [ARM template supports multi-line strings](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/syntax#multi-line-strings), and it's commonly used in resource types such as deployment scripts. Although it can add complexity, I think we do need a more robust solution here.